proc/put.go

```diff
@@ -70,153 +80,133 @@ func (p *Put) maybeWarn(err error) {
+			return nil, err
 		}
 	}
+	return vals, nil
+}
 
-	// Figure out the output descriptor.  We cache it in outmap
-	// but if we haven't seen this input descriptor before or if
-	// the computed type of the expression has changed, recompute it.
-	if !haveDescriptor {
-		origCols := len(in.Type.Columns)
-		cols := make([]zng.Column, origCols, origCols+len(p.clauses))
-		for i, c := range in.Type.Columns {
-			cols[i] = c
-		}
-
-		fields := make([]fieldinfo, len(p.clauses))
-		for k, cl := range p.clauses {
-			newcolumn := zng.Column{
-				Name: cl.target,
-				Type: p.vals[k].Type,
-			}
-
-			position, hasCol := in.Type.ColumnOfField(cl.target)
-			if hasCol {
-				cols[position] = newcolumn
-			} else {
-				position = -1
-				cols = append(cols, newcolumn)
-			}
-			fields[k].valType = p.vals[k].Type
-			fields[k].position = position
-			fields[k].container = p.vals[k].IsContainer()
-		}
-
-		typ, err := p.TypeContext.LookupTypeRecord(cols)
-		if err != nil {
-			p.maybeWarn(err)
-			return in
-		}
-
-		// Compute the order in which to write fields (i.e., if
-		// we're overriding existing fields, write them in the
-		// order they appear in the output record).
-		order := make([]int, len(p.clauses))
-		for k := range order {
-			order[k] = k
+func (p *Put) buildRule(inType *zng.TypeRecord, vals []zng.Value) (*putRule, error) {
+	n := len(inType.Columns)
+	cols := make([]zng.Column, n, n+len(p.clauses))
+	copy(cols, inType.Columns)
+	clauseTypes := make([]clauseType, len(p.clauses))
+	nreplace := 0
+	replace := make([]int, n)
+	for k := range replace {
+		replace[k] = -1
+	}
+	var tail []int
+	for k, cl := range p.clauses {
+		typ := vals[k].Type
+		clauseTypes[k].Type = typ
+		clauseTypes[k].container = zng.IsContainerType(typ)
+		col := zng.Column{Name: cl.target, Type: typ}
+		position, hasCol := inType.ColumnOfField(cl.target)
+		if hasCol {
+			nreplace++
+			replace[position] = k
+			cols[position] = col
+		} else {
+			tail = append(tail, k)
+			cols = append(cols, col)
 		}
-		sort.Slice(order, func(a, b int) bool {
-			if fields[a].position == -1 {
-				return false
-			}
-			if fields[b].position == -1 {
-				return true
-			}
-			return fields[a].position < fields[b].position
-		})
+	}
+	if nreplace == 0 {
+		replace = nil
+	}
+	typ, err := p.TypeContext.LookupTypeRecord(cols)
+	if err != nil {
+		return nil, err
+	}
+	return &putRule{
+		typ:         typ,
+		clauseTypes: clauseTypes,
+		replace:     replace,
+		append:      tail,
+	}, nil
+}
 
-		newinfo := descinfo{
-			typ:    typ,
-			fields: fields,
-			order:  order,
+func clauseTypesMatch(types []clauseType, vals []zng.Value) bool {
+	for k, typ := range types {
+		if vals[k].Type != typ.Type {
+			return false
 		}
+	}
+	return true
+}
 
-		p.outmap[in.Type.ID()] = newinfo
-		descInfo = newinfo
+func (p *Put) lookupRule(inType *zng.TypeRecord, vals []zng.Value) (*putRule, error) {
+	rule := p.rules[inType.ID()]
+	if rule != nil && clauseTypesMatch(rule.clauseTypes, vals) {
```

Comment:
nils can be put in the table on error.  If I make this change, then I'll need to add a check to for error that wasn't otherwise needed.

Reply:
I find `rule ! =nil` clearer in this case anyway because the code requires rule not nil and it doesn't matter if nils are allowed in the table or not for this code to work correctly.

Reply:
Doh. I didn't realize that nil could be in the map.